File: src/back-end/FunChess/src/FunChess.Core.Chess/Pieces/Rook.cs
using FunChess.Core.Chess.Enums;
using FunChess.Core.Chess.Extensions;
using FunChess.Core.Chess.Structs;

namespace FunChess.Core.Chess.Pieces;

internal sealed class Rook : Piece
{
    public override bool Move(Board board, Move move)
    {
        if (!MoveIsValid(board, move, out _)) return false;
        
        board.InternalBoard[move.Next.Index] = board.InternalBoard[move.Previous.Index];
        board.InternalBoard[move.Previous.Index] = Cell.Empty;
        
        CastlingPlay plays = board.Teams[board.Turn].CastlingPlays;
        if ((plays & CastlingPlay.LeftCastling) == CastlingPlay.LeftCastling && move.Previous.Index is 0 or 56)
        {
            plays &= ~CastlingPlay.LeftCastling;
            board.Teams[board.Turn].CastlingPlays = plays;
        }
        else if ((plays & CastlingPlay.RightCastling) == CastlingPlay.RightCastling && move.Previous.Index is 7 or 63)
        {
            plays &= ~CastlingPlay.RightCastling;
            board.Teams[board.Turn].CastlingPlays = plays;
        }
        return true;
    }
    
    public override bool MoveIsValid(Board board, Move move, out SpecialMove specialMove)
    {
        specialMove = SpecialMove.None;
        if (move.DiffX == 0)
        {
            if (Math.Abs(move.DiffY) <= 1)
            {
                return true;
            }
            return move.DiffY < 0 ? board.InternalBoard.CheckSouth(move) : board.InternalBoard.CheckNorth(move);
        }
        if (move.DiffY != 0) return false;
        if (Math.Abs(move.DiffX) <= 1) return true;
        
        return move.DiffX < 0 ? board.InternalBoard.CheckWest(move) : board.InternalBoard.CheckEast(move);
    }
}

File: src/front-end/fun-chess-app/src/core/chess/pieces/Rook.ts
import Piece from "../Piece";
import Board from "../Board";
import CastlingPlay from "../enums/CastlingPlay";
import SpecialMove from "../enums/SpecialMove";
import Move from "../structs/Move";
import { checkEast, checkNorth, checkSouth, checkWest } from "../utils/cells-utils";

class Rook extends Piece {
    private constructor() {
        super(16);
    }

    public static readonly instance: Piece = new Rook();

    public override move(board: Board, move: Move): [boolean, SpecialMove] {
        const tuple = super.move(board, move);
        if (!tuple[0]) return tuple;
        
        let plays = board.teams.get(board.turn)!.castlingPlays;
        if ((plays & CastlingPlay.LeftCastling) === CastlingPlay.LeftCastling && move.previous.index === 0 || move.previous.index === 56) {
            plays &= ~CastlingPlay.LeftCastling;
            board.teams.get(board.turn)!.castlingPlays = plays;
        }
        else if ((plays & CastlingPlay.RightCastling) === CastlingPlay.RightCastling && move.previous.index === 7 || move.previous.index === 63) {
            plays &= ~CastlingPlay.RightCastling;
            board.teams.get(board.turn)!.castlingPlays = plays;
        }
        return tuple;
    }

    public override moveIsValid(board: Board, move: Move): [boolean, SpecialMove] {
        const specialMove = SpecialMove.None;
        if (move.diffX === 0) {
            if (Math.abs(move.diffY) <= 1) {
                return [true, specialMove];
            }
            return [move.diffY < 0 ? checkSouth(board.internalBoard, move) : checkNorth(board.internalBoard, move), specialMove];
        }
        if (move.diffY !== 0) return [false, specialMove];
        if (Math.abs(move.diffX) <= 1) return [true, specialMove];
        
        return [move.diffX < 0 ? checkWest(board.internalBoard, move) : checkEast(board.internalBoard, move), specialMove];
    }
}

export default Rook;